### ISSUE TITLE
proposal: `no_else_after_return`

### ISSUE THREAD

**Issue by goderbauer** [dart-lang/linter#4363]
# no_else_after_return

## Description

Move the code of the else branch below the `if` block.

## Details

When the `if` case always ends with a `return`, the else branch is not necessary and the code can instead be placed below the `if` block. This decreases code complexity and indentation and overall improves readability. 

## Kind

style advice, maybe?

## Bad Examples

```dart
if (foo) {
  // lots of other code...
  return 'foo';
} else {
  return 'not foo';
}
```

## Good Examples 

```dart
if (foo) {
  // lots of other code...
  return 'foo';
}
return 'not foo';
```

## Discussion

Real world example: https://github.com/flutter/flutter/pull/126935#discussion_r1195510913

### Discussion checklist

- [x] List any existing rules this proposal modifies, complements, overlaps or conflicts with.
- [x] List any relevant issues (reported here, the [SDK Tracker], or elsewhere).
- [x] If there's any prior art (e.g., in other linters), please add references here.
- [x] If this proposal corresponds to [Effective Dart] or [Flutter Style Guide] advice, please call it out. (If there isn't any corresponding advice, should there be?)
- [x] If this proposal is motivated by real-world examples, please provide as many details as you can.  Demonstrating potential impact is _especially valuable_.

[Writing Lints]: https://github.com/dart-lang/linter/blob/main/doc/writing-lints.md
[Effective Dart]: https://dart.dev/guides/language/effective-dart
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[SDK Tracker]: https://github.com/dart-lang/sdk/issues

**Comment by srawlins** [comment]
I think typically I would advocate that the former (bad) example is more readable in many cases.

In cases where a return is an early return that allows 100 lines following to be not indented 2 spaces, it makes sense to not use `else`. But often I like to use `else` just to improve readability, like if the then-block is long, or if the condition is long, it ties the state of the following code back to the condition.

Also, as this is a style suggestion, I think we need such a rule to be backed by a style guide, like Effective Dart.

**Comment by bwilkerson** [comment]
@munificent

**Comment by incendial** [comment]
jfyi, if you decide not to add this rule to the linter, it's already available in DCM https://dcm.dev/docs/teams/rules/common/avoid-redundant-else/

**Comment by munificent** [comment]
I think about this every time I have conditional logic that returns on both branches. I tend to lean towards the early return style suggested here, but I'm never fully convinced that it is actually better. I do appreciate the symmetry of having the returns in both branches at the same level.

I think this would be a reasonable lint to support but only if users have to explicitly opt in to enable it if they prefer that style and want consistency. I don't think we'd want "Effective Dart" to have an opinion one way or the other. I just don't think it matters enough either way to be worth bothering users.
